scripts/smb-enum-services.nse

```diff
@@ -0,0 +1,1144 @@
+-- <elem>SERVICE_TYPE_WIN32</elem>
+-- </table>
+-- <elem key="check_point">0</elem>
+-- <elem key="wait_hint">0</elem>
+-- </table>
+-- </table>
+-- <table>
+-- <elem key="displayName">WMI Performance Adapter</elem>
+-- <elem key="serviceName">WmiApSrv</elem>
+-- <table key="serviceStatus">
+-- <table key="state">
+-- <elem>SERVICE_STATE_ACTIVE</elem>
+-- <elem>SERVICE_STATE_ALL</elem>
+-- </table>
+-- <elem key="win32_exit_code">1077</elem>
+-- <elem key="service_exit_code">0</elem>
+-- <table key="controls_accepted">
+-- </table>
+-- <table key="type">
+-- <elem>SERVICE_TYPE_WIN32_OWN_PROCESS</elem>
+-- <elem>SERVICE_TYPE_WIN32</elem>
+-- </table>
+-- <elem key="check_point">0</elem>
+-- <elem key="wait_hint">0</elem>
+-- </table>
+-- </table>
+
+author = "Rewanth Cool"
+license = "Same as Nmap--See https://nmap.org/book/man-legal.html"
+categories = {"discovery","intrusive","safe"}
+
+portrule = shortport.port_or_service({445, 139}, "microsoft-ds", "tcp", "open")
+
+action = function(host, port)
+
+  local status, smbstate = msrpc.start_smb(host, msrpc.SVCCTL_PATH)
+  status, bind_result = msrpc.bind(smbstate, msrpc.SVCCTL_UUID, msrpc.SVCCTL_VERSION, nil)
+
+  if(status == false) then
+    smb.stop(smbstate)
+    return false, bind_result
+  end
+
+  -- Open the service manager
+  stdnse.debug2("Opening the remote service manager")
+
+  status, open_result = msrpc.svcctl_openscmanagerw(smbstate, host.ip)
+
+  if(status == false) then
+    smb.stop(smbstate)
+    return false, open_result
+  end
+
+
+  --@param dwservicetype The type of services to be enumerated.
+  --                     Lookup table for dwservicetype is as follows:
+  --                       SERVICE_DRIVER - 0x0000000B
+  --                       SERVICE_FILE_SYSTEM_DRIVER - 0x00000002
+  --                       SERVICE_KERNEL_DRIVER - 0x00000001
+  --                       SERVICE_WIN32 - 0x00000030
+  --                       SERVICE_WIN32_OWN_PROCESS - 0x00000010 (default)
+  --                       SERVICE_WIN32_SHARE_PROCESS - 0x00000020
+  local dwservicetype = 0x00000010
+
+  --@param dwservicestate The state of the services to be enumerated.
+  --                      Lookup table for dwservicetype is as follows:
+  --                      SERVICE_ACTIVE - 0x00000001
+  --                      SERVICE_INACTIVE - 0x00000002
+  --                      SERVICE_STATE_ALL - 0x00000003 (default)
+  local dwservicestate = 0x00000003
+
+  -- Fetches service name, display name and service status of every service.
+  status, result = msrpc.svcctl_enumservicesstatusw(smbstate, open_result["handle"], dwservicetype, dwservicestate)
+
+  -- Close the service manager
+  stdnse.debug2("Closing the remote service manager")
+
+  status, close_result = msrpc.svcctl_closeservicehandle(smbstate, open_result['handle'])
+
+  if(status == false) then
+    smb.stop(smbstate)
+    return false, close_result
+  end
+
+  smb.stop(smbstate)
+
+  return result
```

Comment:
Here is the recommended format of the result table, with each of the tables being the result of a call to `stdnse.output_table()` to preserve the ordering of the keys:

```
  ALG:
    display_name: Application Layer Gateway Service
    state:
      SERVICE_STATE_RUNNING
    type:
      SERVICE_TYPE_WIN32_OWN_PROCESS
    controls_accepted:
      SERVICE_CONTROL_NETBINDADD
```

Of course I've just truncated the lists for illustration; you would still show all the values for controls, type, and state. Maybe later once we've committed it we can work on better display formats. But this is at least not as large as the current format.